File: algorithms/leetcode/424-longest-repeating-character-replacement/LongestRepeatingCharacterReplacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int left = 0;
        int[] chars = new int[26];
        int maxUnique = 0;
        int result = 0;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            
            chars[ch - 'A']++;
            maxUnique = Math.max(maxUnique, chars[ch - 'A']);
            
            int replacesNeeded = i - left + 1 - maxUnique;
            
            if (replacesNeeded > k) {
                chars[s.charAt(left++) - 'A']--;
            } else {
                result = Math.max(i - left + 1, result);
            }
        }
        return result;
    }
}

File: leetcode/424-longest-repeating-character-replacement/longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
    const chars = {};

    let start = 0;
    let uniqueCount = 0;
    let result = 0;

    for (let i = 0; i < s.length; i++) {
        const ch = s[i];

        chars[ch] = -~chars[ch];
        uniqueCount = Math.max(chars[ch], uniqueCount);

        const nonRepeat = i - start + 1 - uniqueCount;

        if (nonRepeat > k) {
            chars[s[start++]]--;
        } else {
            result = Math.max(result, i - start + 1);
        }
    }

    return result;
};
